File: LeetConsole/Methods/Others/Leet105.cs
using LeetCode.Class;
using System.Linq;

namespace Leet.Methods
{
    /// <summary>
    ///
    /// </summary>
    public class Leet105
    {
        public TreeNode Action()
        {
            //var preorder = new int[] { 3, 9, 20, 15, 7 };
            //var inorder = new int[] { 9, 3, 15, 20, 7 };

            var preorder = new int[] { 1, 2, 3 };
            var inorder = new int[] { 3, 2, 1 };

            var r = BuildTree(preorder, inorder);
            return r;
        }

        public TreeNode BuildTree(int[] preorder, int[] inorder)
        {
            if (preorder.Length == 0) return null;
            var r = new TreeNode(preorder[0]);

            for (int i = 0; i < inorder.Length; i++)
            {
                if (inorder[i] == r.val)
                {
                    if (i > 0)
                    {
                        r.left = BuildTree(preorder.Skip(1).Take(i).ToArray(), inorder.Take(i).ToArray());
                    }
                    if (i < inorder.Length - 1)
                    {
                        r.right = BuildTree(preorder.Skip(i + 1).Take(preorder.Length - i - 1).ToArray(), inorder.Skip(i + 1).Take(preorder.Length - i - 1).ToArray());
                    }
                    break;
                }
            }

            return r;
        }

        public TreeNode BuildTree(int[] preorder, int[] inorder, int ls, int ll, int rs, int rl)
        {
            if (preorder.Length == 0) return null;
            var result = new TreeNode(preorder[0]);

            for (int i = 0; i < inorder.Length; i++)
            {
                if (inorder[i] == result.val)
                {
                    if (i > 0)
                    {
                        result.left = BuildTree(preorder.Skip(1).Take(i).ToArray(), inorder.Take(i).ToArray());
                    }
                    if (i < inorder.Length - 1)
                    {
                        result.right = BuildTree(preorder.Skip(i + 1).Take(preorder.Length - i - 1).ToArray(), inorder.Skip(i + 1).Take(preorder.Length - i - 1).ToArray());
                    }
                    break;
                }
            }

            return result;
        }
    }
}

File: JavaCode/JavaCode/src/main/java/leet/methods/DivideConquer/Leet105.java
package leet.methods.DivideConquer;

import leet._class.Tree.TreeNode;

import java.util.HashMap;
import java.util.Map;

public class Leet105 {

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        int n = preorder.length;
        Map<Integer, Integer> index = new HashMap<>(n); // 预分配空间
        for (int i = 0; i < n; i++) {
            index.put(inorder[i], i);
        }
        return dfs(preorder, 0, n, 0, n, index); // 左闭右开区间
    }

    private TreeNode dfs(int[] preorder, int preL, int preR, int inL, int inR, Map<Integer, Integer> index) {
        if (preL == preR) { // 空节点
            return null;
        }
        int leftSize = index.get(preorder[preL]) - inL; // 左子树的大小
        TreeNode left = dfs(preorder, preL + 1, preL + 1 + leftSize, inL, inL + leftSize, index);
        TreeNode right = dfs(preorder, preL + 1 + leftSize, preR, inL + 1 + leftSize, inR, index);
        return new TreeNode(preorder[preL], left, right);
    }
}
